backend/packages/Upgrade/src/api/services/FeatureFlagService.ts

```diff
@@ -110,36 +140,119 @@ export class FeatureFlagService {
         throw error;
       }
 
-      // creating variations docs
-      const variationDocsToSave =
-        variations &&
-        variations.length > 0 &&
-        variations.map((variation: FlagVariation) => {
-          variation.id = variation.id || uuid();
-          variation.featureFlag = featureFlagDoc;
-          return variation;
-        });
-
-      // saving variations
-      let variationDocs: FlagVariation[];
+      const setSegmentInEx = (inEx: FeatureFlagSegmentExclusion | FeatureFlagSegmentInclusion) => {
+        const segment = inEx.segment;
+        return segment
+          ? {
+              ...featureFlagSegmentInclusion.segment,
+              type: segment.type,
+              userIds: segment.individualForSegment?.map((x) => x.userId) || [],
+              groups:
+                segment.groupForSegment?.map((x) => {
+                  return { type: x.type, groupId: x.groupId };
+                }) || [],
+              subSegmentIds: segment.subSegments?.map((x) => x.id) || [],
+            }
+          : inEx;
+      };
+
+      let includeSegmentExists = true;
```

Comment:
I asked copilot to pull this out 159-223 into own function without the repetition for the segment include and segment exclude. It needs some help with types but I thought something like this looked pretty reasonable, what do you think?

```typescript
async function handleSegment(segmentInclusionExclusion: any, flagDoc: any, flag: any, type: string, logger: any) {

  let segmentExists = true;
  let segmentDoc: Segment;
  let segmentDocToSave: Partial<FeatureFlagSegmentInclusion | FeatureFlagSegmentExclusion> = {};
  if (segmentInclusionExclusion) {
    const segment: any = setSegmentInEx(segmentInclusionExclusion);

    const segmentData: SegmentInputValidator = {
      ...segment,
      id: segment.id || uuid(),
      name: flagDoc.id + ' ' + type + ' Segment',
      description: flagDoc.id + ' ' + type + ' Segment',
      context: flagDoc.context[0],
      type: SEGMENT_TYPE.PRIVATE,
    };
    try {
      segmentDoc = await this.segmentService.upsertSegment(segmentData, logger);
    } catch (err) {
      const error = err as ErrorWithType;
      error.details = 'Error in adding segment in DB';
      error.type = SERVER_ERROR.QUERY_FAILED;
      logger.error(error);
      throw error;
    }
    // creating segment doc
    const tempDoc = type === 'Inclusion' ? new FeatureFlagSegmentInclusion() : new FeatureFlagSegmentExclusion();
    tempDoc.segment = segmentDoc;
    tempDoc.featureFlag = flag;
    segmentDocToSave = this.getSegmentDoc(tempDoc);
  } else {
    segmentExists = false;
  }
  return { segmentExists, segmentDoc, segmentDocToSave };
}
```

Reply:
I think it makes sense to reduce duplication. I'll see what I can do about wrangling the types.